keps/sig-storage/20190129-csi-volume-resizing.md

```diff
@@ -130,6 +130,30 @@ If `NodeExpandVolume` is successful:
 If `NodeExpandVolume` failed:
 - It will add a event to both PVC and Pod about failed resizing and resize operation will be retried.
 
+#### Supporting per-PVC secret refs
+
+To support per-PVC secrets for volume resizing, similar to CSI attach and detach - this proposal expands `CSIPersistentVolumeSource` object to contain `ControllerExpansionSecretRef`:
+
+```
+type CSIPersistentVolumeSource struct {
+    ....
+    // ControllerPublishSecretRef is a reference to the secret object containing
+    // sensitive information to pass to the CSI driver to complete the CSI controller publish
+    ControllerPublishSecretRef *SecretReference
+
+    // ControllerExpansionSecretRef is a reference to secret object containing sensitive
+    // information to pass to the CSI driver to complete CSI controller expansion
+    ControllerExpansionSecretRef *SecretReference
+}
+```
+
+Secrets will be fetched from StorageClass with parameters `csi.storage.k8s.io/controller-expansion-secret-name` and `csi.storage.k8s.io/controller-expansion-secret-namespace`. Resizing secrets will support same templating rules as attach and detach as documented - https://kubernetes-csi.github.io/docs/secrets-and-credentials.html#controller-publishunpublish-secret .
+
+Starting from 1.15 it is expected that all CSI volumes that require secrets for expansion will have `ControllerExpansionSecretRef` field set. If not set
```

Comment:
is the expansion feature in alpha or beta currently?

Reply:
CSI expansion is in alpha currently, but we were hoping to move this to beta in 1.15.

Reply:
the new API fields in stable objects have to remain at alpha for their first release

Reply:
so it sounds like we can't move this feature to beta in 1.15 because moving to beta pretty much requires the field. So our best bet is to introduce the field and make necessary changes but keep CSI resizing in alpha. cc @msau42

Reply:
Can we have resizing secrets go as a separate feature?

Resizing going beta is a blocker for migration going beta and the plugins that we're migrating don't need secrets

Reply:
I think our best bet is to add a new feature gate for CSI resizing secrets. We can merge this feature gate with in general CSI resizing in next release and remove this newly feature gate in next release.

Reply:
that's ugly, but permissible. I'll leave that up to sig-storage folks to decide. document in the API fields that they are alpha level, and which feature gate(s) they depend on

Reply:
We discussed this briefly online. We are going to leave this feature in alpha in 1.15 and introduce the new field.  I will update this KEP.